client/js/contextMenuFactory.js

```diff
@@ -125,6 +125,82 @@ function addKickItem() {
 	});
 }
 
+function addOpItem() {
+	function op(itemData) {
+		socket.emit("input", {
+			target: $("#chat").data("id"),
+			text: "/op " + itemData,
+		});
+	}
+
+	addContextMenuItem({
+		check: (target) =>
+			utils.hasRoleInChannel(target.closest(".chan"), ["op"]) &&
+			!utils.hasRoleInChannel(target.closest(".chan"), ["op"], target.data("name")),
+		className: "action-op",
+		displayName: "Give operator (+o)",
+		data: (target) => target.data("name"),
+		callback: op,
+	});
+}
+
+function addDeopItem() {
+	function deop(itemData) {
+		socket.emit("input", {
+			target: $("#chat").data("id"),
+			text: "/deop " + itemData,
+		});
+	}
+
+	addContextMenuItem({
+		check: (target) =>
+			utils.hasRoleInChannel(target.closest(".chan"), ["op"]) &&
+			utils.hasRoleInChannel(target.closest(".chan"), ["op"], target.data("name")),
+		className: "action-op",
+		displayName: "Revoke operator (-o)",
+		data: (target) => target.data("name"),
+		callback: deop,
+	});
+}
+
+function addVoiceItem() {
+	function voice(itemData) {
+		socket.emit("input", {
+			target: $("#chat").data("id"),
+			text: "/voice " + itemData,
+		});
+	}
+
+	addContextMenuItem({
+		check: (target) =>
+			utils.hasRoleInChannel(target.closest(".chan"), ["op"]) &&
+			!utils.hasRoleInChannel(target.closest(".chan"), ["voice"], target.data("name")),
```

Comment:
I think this will show "Give voice" even if the user is already op. I know that technically users can have both of the user modes, that's how it works, but it seems a bit odd to have it on the UI. The use cases for having both are very small and quite niche.

Thoughts? I don't think it's big enough to delay the PR, but I just thought I'd point it out.

Reply:
The way it shows is correct. Server is aware of all the modes (due to multi mode cap) but they are not propagated to the client.